### PR TITLE
VectorizedArray for ARM NEON

### PR DESCRIPTION
**Pull request by bergbauer**
Finishes the work from #12455

Can somebody with powerful Apple silicon do a full test suite run on this branch? It compiles fine on my laptop and vectorization tests are passing.

### CODE REVIEW

**Review by tamiko — APPROVED** [review]
The CMake part looks good. The rest looks good as well.
Looking forward to seeing how much this improves performance :-)

**Comment by tamiko** [comment]
@tjhei @luca-heltai @ejtovar  *ping* Can you give this a spin?

**Review by masterleinad — APPROVED** [review]
Looks good to me. I can confirm that this passes all vectorization tests with ARM NEON support.

**Comment by tjhei** [comment]
Detection works correctly on my M1: https://ci.tjhei.info/job/dealii-osx/job/PR-15923/2/console

**Comment by tjhei** [comment]
There are a bunch of problems on the ARM ampere build, though: https://ci.tjhei.info/job/dealii-ampere/job/PR-15923/2/console

**Review by peterrum — COMMENTED** [review]
Nice :1st_place_medal:

**Comment by peterrum** [comment]
@bergbauer Could you also add a change-log entry.

File: include/deal.II/base/vectorization.h [hunk 1]
```diff
@@ -69,6 +69,8 @@
 #    undef vector
 #    undef pixel
 #    undef bool
+#  elif defined(__ARM_NEON__)
+#    include <arm_neon.h>
 #  else
 #    include <x86intrin.h>
 #  endif
```
Comment (bergbauer) [inline include/deal.II/base/vectorization.h:76]:
```
In file included from /home/ubuntu/jenkins/workspace/dealii-ampere_PR-15923/include/deal.II/base/vectorization.h:75:
In file included from /usr/lib/llvm-15/lib/clang/15.0.7/include/x86intrin.h:15:
/usr/lib/llvm-15/lib/clang/15.0.7/include/immintrin.h:14:2: error: "This header is only meant to be used on x86 and x64 architecture"
#error "This header is only meant to be used on x86 and x64 architecture"
```
Apparently `__ARM_NEON__` is not defined for Ampere which is why it includes the wrong header. Is there a different define for Ampere?

**Comment by kronbichler** [comment]
The error starts with
```
In file included from /home/ubuntu/jenkins/workspace/dealii-ampere_PR-15923/include/deal.II/base/vectorization.h:75:
In file included from /usr/lib/llvm-15/lib/clang/15.0.7/include/x86intrin.h:13:
/usr/lib/llvm-15/lib/clang/15.0.7/include/ia32intrin.h:200:10: error: use of undeclared identifier '__builtin_ia32_readeflags_u32'
  return __builtin_ia32_readeflags_u32();
         ^
```
This comes from https://github.com/dealii/dealii/blob/26c7c449117cffffd4f6131264e8dcc5d780f5a1/include/deal.II/base/vectorization.h#L72-L74 Your code already includes a suitable define flag, but I wonder if your define flag `#  elif defined(__ARM_NEON__)` is correct, with the end resulting being that we instead include the x86 intrinsics on ampere. Can you try if the more generic flag
```c++
#elif defined(__aarch64__)
#include <arm_neon.h>
```
works both on your machine and ampere? As far as I can see from https://sourceforge.net/p/predef/wiki/Architectures/ this looks to be the correct flag (Neon is also on 32-bit arm architectures, but I don't think we need to care for those)? Also, to improve safety I suggest to wrap the `#include <x86intrin.h>` file into an
```c++
#elif defined(__x86_64__)
#include <x86intrin.h>
```
I again ignore the 32-bit x86/386/686 architectures, which I think is OK in 2023.

File: include/deal.II/base/vectorization.h [hunk 2]
```diff
@@ -966,6 +968,565 @@ vectorized_transpose_and_store(const bool                            add_into,
 
 #ifndef DOXYGEN
 
+#  if DEAL_II_VECTORIZATION_WIDTH_IN_BITS >= 128 && defined(__ARM_NEON__)
```
Comment (kronbichler) [inline include/deal.II/base/vectorization.h:971]: Whichever preprocessor flag we settle for regading the header should also be set here.

**Review by kronbichler — COMMENTED** [review]
Looks good apart from the preprocessor define flag.

**Comment by kronbichler** [comment]
`__ARM_NEON` seems to work on `ARM64 gcc`, `ARM gcc` and `ARM 64-bit clang` according to godbolt, :+1:

File: include/deal.II/base/numbers.h [hunk 3]
```diff
@@ -130,6 +130,8 @@ namespace internal
       8;
 #elif DEAL_II_VECTORIZATION_WIDTH_IN_BITS >= 128 && defined(__SSE2__)
       4;
+#elif DEAL_II_VECTORIZATION_WIDTH_IN_BITS >= 128 && defined(__ARM_NEON__)
```
Comment (kronbichler) [inline include/deal.II/base/numbers.h:133]:
```suggestion
#elif DEAL_II_VECTORIZATION_WIDTH_IN_BITS >= 128 && defined(__ARM_NEON)
```

File: include/deal.II/base/vectorization.h [hunk 4]
```diff
@@ -5636,6 +6196,89 @@ compare_and_apply_mask(const VectorizedArray<double, 2> &left,
   return result;
 }
 
+#  endif
+
+#  if DEAL_II_VECTORIZATION_WIDTH_IN_BITS >= 128 && defined(__ARM_NEON__)
```
Comment (kronbichler) [inline include/deal.II/base/vectorization.h:6201]:
```suggestion
#  if DEAL_II_VECTORIZATION_WIDTH_IN_BITS >= 128 && defined(__ARM_NEON)
```

**Review by kronbichler — APPROVED** [review]
Looks good, if you change two more places.

File: cmake/checks/check_01_cpu_features.cmake [hunk 5]
```diff
@@ -245,6 +246,37 @@ if(DEAL_II_ALLOW_PLATFORM_INTROSPECTION)
     "
     DEAL_II_HAVE_ALTIVEC)
 
+    CHECK_CXX_SOURCE_RUNS(
+      "
+      #include <arm_neon.h>
```
Comment (kronbichler) [inline cmake/checks/check_01_cpu_features.cmake:251]: If we wanted to be really sure, we would add
```suggestion
      #ifndef __ARM_NEON
      #error Preprocessor flag not found
      #endif
      #include <arm_neon.h>
```

**Review by tjhei — COMMENTED** [review]
Great, things are working now!

File: include/deal.II/base/vectorization.h [hunk 6]
```diff
@@ -966,6 +968,565 @@ vectorized_transpose_and_store(const bool                            add_into,
 
 #ifndef DOXYGEN
 
+#  if DEAL_II_VECTORIZATION_WIDTH_IN_BITS >= 128 && defined(__ARM_NEON)
```
[the remark this answers is not in the document]
Comment (bergbauer) [reply include/deal.II/base/vectorization.h]: I made the define available in `config.h.in`. Done.

**Pull request merged**